orttraining/orttraining/test/graph/gradient_graph_builder_test.cc

```diff
@@ -997,11 +997,115 @@ class PipelineBatchPlanner {
+  Node** forward_record_after_send,
+  Node** backward_wait_before_recv,
+  Node** backward_wait_after_recv,
+  Node** backward_record_before_send,
+  Node** backward_record_after_send) {
+  // Initialize retrieved nodes.
+  // Non-existing nodes may hold NULL forever.
+  // Existing nodes may get valid pointers below.
+  *forward_wait_before_recv = nullptr;
+  *forward_wait_after_recv = nullptr;
+  *forward_record_before_send = nullptr;
+  *forward_record_after_send = nullptr;
+  *backward_wait_before_recv = nullptr;
+  *backward_wait_after_recv = nullptr;
+  *backward_record_before_send = nullptr;
+  *backward_record_after_send = nullptr;
+
+  // Declare container for WaitEvent's in topological order.
+  std::vector<Node*> waits;
+  // Declare container for RecordEvent's in topological order.
+  std::vector<Node*> records;
+
+  // Find out WaitEvent's and RecordEvent's.
+  GraphViewer graph_viewer(graph);
+  for (auto& node_idx : graph_viewer.GetNodesInTopologicalOrder()) {
+    Node* node = graph.GetNode(node_idx);
+    if (node->OpType() == "WaitEvent") {
+      waits.push_back(node);
+    } else if (node->OpType() == "RecordEvent") {
+      records.push_back(node);
+    }
+  }
+
+  if (waits.size() == size_t(4)) {
+    // Each of first stage and middle stages has 4 WaitEvent's.
+    *forward_wait_before_recv = waits[0];
+    *forward_wait_after_recv = waits[1];
+    *backward_wait_before_recv = waits[2];
+    *backward_wait_after_recv = waits[3];
+  } else if (waits.size() == size_t(2)) {
+    // Last stage has 2 WaitEvent's.
+    *forward_wait_before_recv = waits[0];
+    *forward_wait_after_recv = waits[1];
+  } else {
+    ORT_THROW("Wrong number of WaitEvent operators: ", waits.size(), "Expected value is either 2 or 4.");
+  }
+
+  if (records.size() == size_t(4)) {
+    // Each of first stage and middle stages has 4 RecordEvent's.
+    *forward_record_before_send = records[0];
+    *forward_record_after_send = records[1];
+    *backward_record_before_send = records[2];
+    *backward_record_after_send = records[3];
+  } else if (waits.size() == size_t(2)) {
+    // Last stage has 2 RecordEvent's.
+    *backward_record_before_send = records[0];
+    *backward_record_after_send = records[1];
+  } else {
+    ORT_THROW("Wrong number of RecordEvent operators: ", waits.size(), ". Expected value is either 2 or 4.");
+  }
+}
+
+void RetrieveSendRecvOperators(
+  Graph& graph, 
+  Node** forward_recv,
+  Node** forward_send,
+  Node** backward_recv,
+  Node** backward_send) {
+  // Initialize retrieved nodes.
+  // Non-existing nodes may hold NULL forever.
+  // Existing nodes may get valid pointers below.
+  *forward_recv = nullptr;
+  *forward_send = nullptr;
+  *backward_recv = nullptr;
+  *backward_send = nullptr;
+
+  auto is_backward = [](Node& node) {
+    return (node.Description() == "Backward pass");
+  };
+
+  // Search for Send's and Recv's by assuming that
+  // there are only one Send and one Recv in forward/backward.
```

Comment:
>// there are only one Send and one Recv in forward/backward. [](start = 2, length = 60)

Could you enforce this uniqueness in the loop below with a check before assignment?

Reply:
Sure. Will do it in my next PR. We will have

```c
      if (is_backward(node)) {
        // backward_send can only be assigned one valid pointer.
        // If it is assigned more than once, it means we have multiple
        // Send in backward pass and therefore our assumption doesn't hold. 
        // This check ensure that only we only update *backward_send when
        // its value is NULL and guards our one-Recv assumption. 
        ASSERT_TRUE(!(*backward_send));
        *backward_send = &node;
      } else {
        // Guard the uniqueness of Send in the forward pass by throwing
        // when *forward_send already carries a valid pointer.
        ASSERT_TRUE(!(*forward_send));
        *forward_send = &node;
      }
    } else if (node.OpType() == "Recv") {
      if (is_backward(node)) {
        // Guard the uniqueness of Recv in the backward pass by throwing
        // when *backward_recv already carries a valid pointer.
        ASSERT_TRUE(!(*backward_recv));
        *backward_recv = &node;
      } else {
        // Guard the uniqueness of Recv in the forwaard pass by throwing
        // when *forward_recv already carries a valid pointer.
        *forward_recv = &node;
      }
    }
```

---
In reply to: [426074045](https://github.com/microsoft/onnxruntime/pull/3823#discussion_r426074045) [](ancestors = 426074045)